pkg/kubelet/server/server.go

```diff
@@ -826,7 +826,23 @@ func (s *Server) getRun(request *restful.Request, response *restful.Response) {
 
 	// For legacy reasons, run uses different query param than exec.
 	params.cmd = strings.Split(request.QueryParameter("cmd"), " ")
-	data, err := s.host.RunInContainer(kubecontainer.GetPodFullName(pod), params.podUID, params.containerName, params.cmd)
+
+	var timeout time.Duration
+	t := request.QueryParameter("timeoutSeconds")
```

Comment:
rather than adding a query parameter, can we use the deadline from the request context?

Reply:
> rather than adding a query parameter, can we use the deadline from the request context?

@tallclair  thanks for the reviews. yes, using the deadline from the request context could complete the request.

I have a small question: we still pass the `timeout` as 0 in `kl.runner.RunInContainer`? then, is there a chance that client cancel the request, and server finish this request (due to request.ctx), but the cmd is still running in container, since the cancel message can not pass into container.
```go
// RunInContainer runs a command in a container, returns the combined stdout, stderr as an array of bytes
func (kl *Kubelet) RunInContainer(podFullName string, podUID types.UID, containerName string, cmd []string) ([]byte, error) {
	container, err := kl.findContainer(podFullName, podUID, containerName)
	if err != nil {
		return nil, err
	}
	if container == nil {
		return nil, fmt.Errorf("container not found (%q)", containerName)
	}
	// TODO(tallclair): Pass a proper timeout value.
	return kl.runner.RunInContainer(container.ID, cmd, 0)
}
```

Reply:
@tallclair 
Or, we pass the `ctx` instead of `timeout` in func `ExecSync`, and move the codes about composing ctx out of func `ExecSync` to where we use this func. In this way, we could pass the client cancel message into container smoothly? 
```
// ExecSync executes a command in the container, and returns the stdout output.
// If command exits with a non-zero exit code, an error is returned.
func (r *remoteRuntimeService) ExecSync(containerID string, cmd []string, timeout time.Duration) (stdout []byte, stderr []byte, err error) {
	klog.V(10).InfoS("[RemoteRuntimeService] ExecSync", "containerID", containerID, "timeout", timeout)
	// Do not set timeout when timeout is 0.
	var ctx context.Context
	var cancel context.CancelFunc
	if timeout != 0 {
		// Use timeout + default timeout (2 minutes) as timeout to leave some time for
		// the runtime to do cleanup.
		ctx, cancel = getContextWithTimeout(r.timeout + timeout)
	} else {
		ctx, cancel = getContextWithCancel()
	}
	defer cancel()

	if r.useV1API() {
		return r.execSyncV1(ctx, containerID, cmd, timeout)
	}

	return r.execSyncV1alpha2(ctx, containerID, cmd, timeout)
}
```